reactor-core/src/main/java/reactor/core/publisher/FluxBufferPredicate.java

```diff
@@ -156,11 +164,11 @@ public void request(long n) {
 					// we'll continue requesting one by one)
 					if (!DrainUtils.postCompleteRequest(n,
 							actual,
-							this,
-							REQUESTED,
+							this, REQUESTED_BUFFERS,
 							this,
 							this)) {
-						s.request(1);
+						Operators.addCap(REQUESTED_FROM_SOURCE, this, n);
+						s.request(n);
```

Comment:
@simonbasle I have one question about that you said before  
https://github.com/reactor/reactor-core/pull/2027#discussion_r374182115
> a Subscriber is responsible for ensuring the upstream requests it makes are not done in parallel (rule 2.7) so that line MUST NOT be executed at the same time

as far as I understand `s.request()` could be invoked from `BufferPredicateSubscriber#request()` and from `BufferPredicateSubscriber#onNext()` and they could be invoked by different `Thread`s.  
 How does `rule 2.7` is satisfied here or it is not applicable here?

Reply:
mmh with the `REQUESTED_FROM_SOURCE` guard, but we indeed don't really use that to ensure `n` is correct here. let's unpack that a little:
 - `addCap` should lose to `compareAndSet` (other writes "happens before")
 - since parallel requests are illegal, the only way addCap loses is if `compareAndSet` switched from 0 to 1, so at that point `requestedFromSource` == `0` or `1`
 - it should thus be safe to use the return value of `addCap` (0 or 1) as a modifier to `n` for the parent request:

```java
long delta = Operators.addCap(REQUESTED_FROM_SOURCE, this, n);
s.request(n - delta);
```

Reply:
the above doesn't seem to be necessary. I have added tests around racing `request` and `onNext` and validated the current code doesn't lead to backpressure errors.